### PR TITLE
model: support solar

### PR DESCRIPTION
**Pull request by ppraneth**
## SGLang vs. vLLM: Verified MMLU Benchmark on SOLAR-10.7B Model

| Model | Metric | SGLang | vLLM | SGLang Advantage |
| :--- | :--- | :--- | :--- | :--- |
| **upstage/SOLAR-10.7B-Instruct-v1.0** | Avg. MMLU Accuracy | 65.2% | 65.2% | 0.0% |
| | Total Latency | 1,056.1s | 6,972.9s | **6.60x Faster** |

## **Comparison with Official MMLU Scores**

**For `upstage/SOLAR-10.7B-Instruct-v1.0`:**
* Official MMLU Score: **66.2%**
* SGLang/vLLM results were **1.0% lower** than the official score.

-----

### SOLAR-10.7B Benchmark Results

| Metric | SGLang | vLLM | SGLang vs. vLLM Difference |
| :--- | :--- | :--- | :--- |
| **Average Accuracy** | 65.2% | 65.2% | **0.0%** |
| **Total Latency** | 1,056.138s | 6,972.980s | **-5916.842s (6.60x faster)** |

#### Detailed Subject-by-Subject Comparison (SOLAR-10.7B)

| Subject | Questions | SGLang Acc | vLLM Acc | SGLang vs. vLLM Diff |
| :--- | :--- | :--- | :--- | :--- |
| Abstract Algebra | 100 | 40.0% | 40.0% | 0.0% |
| Anatomy | 135 | 61.5% | 61.5% | 0.0% |
| Astronomy | 152 | 73.7% | 73.7% | 0.0% |
| Business Ethics | 100 | 74.0% | 74.0% | 0.0% |
| Clinical Knowledge | 265 | 67.9% | 67.9% | 0.0% |
| College Biology | 144 | 75.7% | 75.7% | 0.0% |
| College Chemistry | 100 | 44.0% | 44.0% | 0.0% |
| College Computer Science | 100 | 52.0% | 52.0% | 0.0% |
| College Mathematics | 100 | 28.0% | 28.0% | 0.0% |
| College Medicine | 173 | 66.5% | 66.5% | 0.0% |
| College Physics | 102 | 38.2% | 38.2% | 0.0% |
| Computer Security | 100 | 76.0% | 76.0% | 0.0% |
| Conceptual Physics | 235 | 60.9% | 61.3% | **-0.4%** |
| Econometrics | 114 | 50.0% | 50.0% | 0.0% |
| Electrical Engineering | 145 | 64.1% | 64.1% | 0.0% |
| Elementary Mathematics | 378 | 47.9% | 47.9% | 0.0% |
| Formal Logic | 126 | 46.0% | 46.0% | 0.0% |
| Global Facts | 100 | 35.0% | 35.0% | 0.0% |
| High School Biology | 310 | 80.0% | 80.0% | 0.0% |
| High School Chemistry | 203 | 51.7% | 51.2% | **+0.5%** |
| High School Computer Science | 100 | 72.0% | 72.0% | 0.0% |
| High School European History | 165 | 79.4% | 80.0% | **-0.6%** |
| High School Geography | 198 | 86.9% | 86.9% | 0.0% |
| High School Government and Politics | 193 | 90.7% | 90.7% | 0.0% |
| High School Macroeconomics | 390 | 66.2% | 66.2% | 0.0% |
| High School Mathematics | 270 | 38.1% | 38.5% | **-0.4%** |
| High School Microeconomics | 238 | 71.8% | 71.8% | 0.0% |
| High School Physics | 151 | 37.1% | 37.1% | 0.0% |
| High School Psychology | 545 | 84.8% | 84.6% | **+0.2%** |
| High School Statistics | 216 | 55.1% | 55.6% | **-0.5%** |
| High School US History | 204 | 85.3% | 85.8% | **-0.5%** |
| High School World History | 237 | 85.7% | 85.7% | 0.0% |
| Human Aging | 223 | 68.2% | 68.2% | 0.0% |
| Human Sexuality | 131 | 74.8% | 74.8% | 0.0% |
| International Law | 121 | 77.7% | 77.7% | 0.0% |
| Jurisprudence | 108 | 80.6% | 80.6% | 0.0% |
| Logical Fallacies | 163 | 75.5% | 75.5% | 0.0% |
| Machine Learning | 112 | 44.6% | 44.6% | 0.0% |
| Management | 103 | 82.5% | 82.5% | 0.0% |
| Marketing | 234 | 86.3% | 86.3% | 0.0% |
| Medical Genetics | 100 | 70.0% | 69.0% | **+1.0%** |
| Miscellaneous | 783 | 80.2% | 80.1% | **+0.1%** |
| Moral Disputes | 346 | 75.4% | 75.7% | **-0.3%** |
| Moral Scenarios | 895 | 39.3% | 39.2% | **+0.1%** |
| Nutrition | 306 | 76.1% | 75.8% | **+0.3%** |
| Philosophy | 311 | 72.3% | 72.7% | **-0.4%** |
| Prehistory | 324 | 78.7% | 78.7% | 0.0% |
| Professional Accounting | 282 | 49.6% | 50.4% | **-0.8%** |
| Professional Law | 1534 | 49.5% | 49.6% | **-0.1%** |
| Professional Medicine | 272 | 73.5% | 74.3% | **-0.8%** |
| Professional Psychology | 612 | 69.0% | 69.0% | 0.0% |
| Public Relations | 110 | 69.1% | 69.1% | 0.0% |
| Security Studies | 245 | 73.5% | 73.5% | 0.0% |
| Sociology | 201 | 84.1% | 84.1% | 0.0% |
| US Foreign Policy | 100 | 90.0% | 90.0% | 0.0% |
| Virology | 166 | 58.4% | 58.4% | 0.0% |
| World Religions | 171 | 78.4% | 78.4% | 0.0% |

-----

#### SGLang Benchmark Commands

```bash
# For SOLAR-10.7B
!python -m sglang.launch_server \
  --model-path upstage/SOLAR-10.7B-Instruct-v1.0 \
  --tensor-parallel-size 2 \
  --port 30000
!python3 bench_sglang.py 
```

#### Vllm Benchmark Commands
```bash
# For SOLAR-10.7B
!python3 -m vllm.entrypoints.api_server \
  --tokenizer-mode auto \
  --model upstage/SOLAR-10.7B-Instruct-v1.0 \
  --tensor-parallel-size 2 \
  --disable-log-requests \
  --port 21000
!python3 bench_other.py  --backend vllm
```

## Motivation
To add Solar LLM
## Checklist

- [X] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [X] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [X] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [X] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Review by mickqian — APPROVED** [review]

**Comment by mickqian** [comment]
For such a huge speed difference with vLLM, could it be that the launch command is not optimal?

File: python/sglang/srt/models/solar.py [hunk 1]
```diff
@@ -0,0 +1,482 @@
+from collections.abc import Iterable
```
Comment (mickqian) [inline python/sglang/srt/models/solar.py:26]: Add copyright at here

Comment (ppraneth) [reply python/sglang/srt/models/solar.py:26]: Done

**Comment by ppraneth** [comment]
> For such a huge speed difference with vLLM, could it be that the launch command is not optimal?

Probably as i used kaggle for testing
for sglang:
```
import threading
import subprocess
import time

def launch_server():
    process = subprocess.Popen(
        ["python", "-m", "sglang.launch_server", "--model-path", "upstage/SOLAR-10.7B-Instruct-v1.0","--tensor-parallel-size","2","--port", "30000"],
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True
    )
    for line in process.stdout:
        print("[SERVER]", line.strip())

# Start server thread
server_thread = threading.Thread(target=launch_server)
server_thread.start()
```
```
!python3 bench_sglang.py 

```

For vllm:
```
import threading
import subprocess
import time

def launch_server():
    process = subprocess.Popen(
        ["python", "-m", "vllm.entrypoints.api_server","--tokenizer-mode","auto","--model", "upstage/SOLAR-10.7B-Instruct-v1.0","--tensor-parallel-size","2","--disable-log-requests","--port", "21000"],
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True
    )
    for line in process.stdout:
        print("[SERVER]", line.strip())

# Start server thread
server_thread = threading.Thread(target=launch_server)
server_thread.start()
```
```
!python3 bench_other.py --backend vllm
```

File: python/sglang/srt/models/solar.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+# limitations under the License.
+# Adapted from https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/solar.py
+from collections.abc import Iterable
+from typing import Any, List, Optional, Tuple, Union
+
+import torch
+from torch import nn
+from transformers import PretrainedConfig
+
+from sglang.python.sglang.srt.distributed.parallel_state import (
```
Comment (zhyncs) [inline python/sglang/srt/models/solar.py:33]:
```
Ignore import error when loading sglang.srt.models.solar: No module named 'sglang.python'
```

@hnyls2002

Comment (hnyls2002) [reply python/sglang/srt/models/solar.py:33]: @mickqian

Comment (ppraneth) [reply python/sglang/srt/models/solar.py:33]: Just need to remove sglang.python
